community/kernel/src/main/java/org/neo4j/kernel/impl/query/ResultBuffer.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (c) 2002-2018 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.query;
+
+import org.neo4j.values.AnyValue;
+
+/**
+ * Result buffer holding the results of a query execution.
+ */
+public interface ResultBuffer
+{
+    /**
+     * Signal that a new query execution is starting, which will return {@code valuesPerResult}  many values per result.
+     */
+    void beginQueryExecution( int valuesPerResult );
+
+    /**
+     * Prepare the result stage for the next result row.
+     *
+     * @return the id of the new row, or -1 if the stage could not be cleared.
+     */
+    long beginResultRow() throws ResultBufferException;
```

Comment:
Do we need blocking and non-blocking versions of this method, or is the caller responsible for potentially blocking? If the caller is responsible, how can it be notified without having to poll this result?

Reply:
Yeah, that might be needed eventually.

Currently this methods blocks in Bolt and always back-pressures in Core API. I think we can wait with threading concerns until we have the basic interface in place with the current 1-thread model.